src/main/java/org/zalando/nakadi/service/subscription/SubscriptionTimeLagService.java

```diff
@@ -0,0 +1,129 @@
+    private static final int COMPLETE_TIMEOUT_MS = 30000;
+    private static final int LAG_CALCULATION_PARALLELISM = 10;
+
+    private final TimelineService timelineService;
+    private final NakadiCursorComparator cursorComparator;
+
+    @Autowired
+    public SubscriptionTimeLagService(final TimelineService timelineService,
+                                      final NakadiCursorComparator cursorComparator) {
+        this.timelineService = timelineService;
+        this.cursorComparator = cursorComparator;
+    }
+
+    public Map<EventTypePartition, Duration> getTimeLags(final Collection<NakadiCursor> committedPositions,
+                                                         final List<PartitionEndStatistics> endPositions)
+            throws ErrorGettingCursorTimeLagException, InconsistentStateException {
+
+        final ExecutorService executor = Executors.newFixedThreadPool(LAG_CALCULATION_PARALLELISM);
+
+        final Map<EventTypePartition, Future<Duration>> futures = new HashMap<>();
+        final Map<EventTypePartition, Duration> result = new HashMap<>();
+
+        for (final NakadiCursor cursor : committedPositions) {
+            final EventTypePartition partition = new EventTypePartition(cursor.getTimeline().getEventType(),
+                    cursor.getPartition());
+            if (isCursorAtTail(cursor, endPositions)) {
+                result.put(partition, Duration.ZERO);
+            } else {
+                futures.put(partition, executor.submit(() -> getNextEventTimeLag(cursor)));
+            }
+        }
+
+        executor.shutdown();
+        try {
+            final boolean finished = executor.awaitTermination(COMPLETE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!finished) {
+                throw new InconsistentStateException("Timeout occurred when getting subscription time lag");
+            }
+            for (final EventTypePartition partition : futures.keySet()) {
+                final Future<Duration> future = futures.get(partition);
+                try {
+                    result.put(partition, future.get());
+                } catch (final ExecutionException e) {
+                    throw e.getCause();
+                }
+            }
+            return result;
+
+        } catch (final InconsistentStateException | ErrorGettingCursorTimeLagException e) {
+            throw e;
+        } catch (final Throwable e) {
+            throw new InconsistentStateException("Unexpected error occurred when getting subscription time lag", e);
+        }
+    }
+
+    private Duration getNextEventTimeLag(final NakadiCursor cursor) throws ErrorGettingCursorTimeLagException,
+            InconsistentStateException {
+        try {
+            final EventConsumer consumer = timelineService.createEventConsumer(
+                    "time-lag-checker-" + UUID.randomUUID().toString(), ImmutableList.of(cursor));
+
+            final ConsumedEvent nextEvent = executeWithRetry(
```

Comment:
I thought this execute with retries was exclusively used by tests. Would you consider adding a parameter to `readEvents` to pass the amount of time you are willing to wait? Then you give control to that method and let it use Kafka to manage it. I think that KafkaConsumer accepts some parameters that define precisely "fetch_wait_time_ms". You could use it, when building a new KafkaConsumer. Then you just need to pool and read whatever comes with it.

Reply:
I think I was just lazy to do all that refactoring to be able to pass that parameter :) 
I will do it

Reply:
@v-stepanov I think it might be there already as the last parameter in the constructor. It's called "pollTimeout"